### PR TITLE
example 4: visualizing information

### PR DESCRIPTION
**Pull request by onewhl**
Implemented `Visualizing information` use case.
Added a tool window that shows the basic statistics of classes in the project: 
`class name`
`field number`
`method number`
`lines of code`
`number of children`
Screenshot from 2021-08-31 18-07-47

### CODE REVIEW

File: demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/util/PsiUtil.kt [hunk 1]
```diff
@@ -33,3 +38,35 @@ fun concatFiltered(
     array2.filterTo(filteredList, condition)
     return filteredList.toTypedArray()
 }
+
+/*
+    Extracts all Java classes from the project.
+ */
+fun extractClasses(project: Project?): MutableList<PsiClass> {
```
Comment (vovak) [inline demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/util/PsiUtil.kt]: extractJavaClasses?

Comment (vovak) [inline demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/util/PsiUtil.kt]: Should `project` really be nullable? It's not nullable in the only usage in this branch.

File: demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/util/PsiUtil.kt [hunk 2]
```diff
@@ -33,3 +38,35 @@ fun concatFiltered(
...
+    val classes: MutableList<PsiClass> = ArrayList()
+
+    ProjectFileIndex.SERVICE.getInstance(project).iterateContent { file: VirtualFile? ->
```
Comment (vovak) [inline demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/util/PsiUtil.kt]: Could the iteration code be more concise?
https://github.com/JetBrains-Research/refactoring-workshop-demo/pull/3#discussion_r700453229

File: demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/util/PsiUtil.kt [hunk 3]
```diff
@@ -33,3 +38,35 @@ fun concatFiltered(
...
+        }
+        true
+    }
+    return classes
+}
+
+/*
+    Calculates the number of lines in the text.
+ */
+fun countLines(clazzText: String): Int {
```
Comment (vovak) [inline demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/util/PsiUtil.kt]: `fun countLines(classText: String) = classText.toCharArray().count{it == '\n' || it == '\r'}`

File: demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/statistics/MyToolWindowContent.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        val classes = extractClasses(project)
+        classes.forEach { clazz ->
+            val statistics =
+                ClassStatistics(
+                    clazz.fields.size,
+                    clazz.methods.size,
+                    countLines(clazz.text),
+                    clazz.children.size
+                )
+            results[clazz.qualifiedName!!] = statistics
```
Comment (vovak) [inline demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/statistics/MyToolWindowContent.kt]: what happens if `clazz.qualifiedName == null?`

File: demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/statistics/ClassStatistics.kt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package org.jetbrains.research.refactoringDemoPlugin.statistics
+
+data class ClassStatistics(
+    // number of methods in the class
+    val methodNumber: Int,
```
Comment (vovak) [inline demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/statistics/ClassStatistics.kt]: methodCount

File: demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/statistics/ClassStatistics.kt [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+    // number of fields in the class
+    val fieldsNumber: Int,
+    // lines of code inside the class
+    val loc: Int,
+    // number of children (subclasses)
+    val noc: Int
```
Comment (vovak) [inline demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/statistics/ClassStatistics.kt]: subclassCount

File: demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/statistics/ClassStatistics.kt [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package org.jetbrains.research.refactoringDemoPlugin.statistics
+
+data class ClassStatistics(
+    // number of methods in the class
+    val methodNumber: Int,
+    // number of fields in the class
+    val fieldsNumber: Int,
```
Comment (vovak) [inline demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/statistics/ClassStatistics.kt]: ditto

**Review by vovak — CHANGES_REQUESTED** [review]
Thank you!
Would it be hard to filter out the lines containing only formatting and comments, thus calculating LLOC in addition to LOC? It would also serve as a slightly more detailed example of handling file contents at the lower level.

File: demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/statistics/MyToolWindowContent.kt (hunk 4 above)
Comment (onewhl) [reply demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/statistics/MyToolWindowContent.kt]: added a check if ` clazz.qualifiedName` is not null ([line 49](https://github.com/JetBrains-Research/refactoring-workshop-demo/pull/4/files#diff-90881d896fa0ae48fa25f35270705453d2f6b91eb9e005e05b37b47ac35fe545R49))
so, we start calculating the statistics only if `clazz.qualifiedName` is not null, but Kotlin requires to use !! even we have already made sure that `clazz.qualifiedName` was not null

File: demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/util/PsiUtil.kt (hunk 2 above)
Comment (onewhl) [reply demo-plugin/src/main/kotlin/org/jetbrains/research/refactoringDemoPlugin/util/PsiUtil.kt]: Done.

**Pull request merged**